src/Storages/Distributed/DirectoryMonitor.cpp

```diff
@@ -373,6 +374,13 @@ StorageDistributedDirectoryMonitor::~StorageDistributedDirectoryMonitor()
     }
 }
 
+void StorageDistributedDirectoryMonitor::requestUpdatePool()
+{
+    const char * user_pw_end = strchr(name.data(), '@');
+    if (!user_pw_end && startsWith(name, "shard"))
```

Comment:
I don't think that re-creating pool is enough, since changing the cluster definition can change the destination server, but you may still have pending blocks.

So suppose you have the following cluster:

```xml
<remote_servers>
  <main_cluster>
    <shard>
      <internal_replication>true</internal_replication>
      <replica>
        <host>foo-1</host>
        <port>9000</port>
      </replica>
      <replica>
        <host>foo-2</host>
        <port>9000</port>
      </replica>
    </shard>
    <shard>
      <internal_replication>true</internal_replication>
      <replica>
        <host>bar-1</host>
        <port>9000</port>
      </replica>
      <replica>
        <host>bar-2</host>
        <port>9000</port>
      </replica>
    </shard>
  </main_cluster>
</remote_servers>
```

And the `foo*` shard removed from the cluster definition.
In this case `shard1` will became `bar*` not `foo*`, but all pending blocks should be sent to `foo*` not `bar*`, and during this time any INSERT should be blocked since you cannot write new blocks until you sync old one.
And also there are some trickery with `internal_replication=false`.

So that said that it is not that simple, and it looks to me that it is better to re-create DirectryMonitor objects, plus it requires good integration test, that will cover all possible variants.